wgengine/magicsock/magicsock.go

```diff
@@ -4656,6 +4688,11 @@ func (de *endpoint) updateFromNode(n *tailcfg.Node, heartbeatDisabled bool) {
 	de.sendFunc.Store(nil)
 }
 
+// supportsDisco reports whether or not the endpoint supports disco.
+func (de *endpoint) supportsDisco() bool {
+	return !de.discoKey.IsZero()
```

Comment:
Can we use IsWireGuardOnly?

Reply:
pre 0.100 nodes match this, but do not match IsWireGuardOnly. While that shouldn't be a problem due to guards elsewhere in the code, should those guards change, this check is the safer one, as it guards against entering a code path that will panic the client (the noise shared key production from discoKey). As such I think being specific here is more robust to future change.

Reply:
I looked at this again briefly. I can see a few issues here:

- If the compiler fails to optimize zero checks well, the 32 byte zero check could be expensive, though I beleive this isn't in particularly hot paths. I swapped out for an empty string check against discoShort, which assuming the commentary on the struct remains correct should be functionally equivalent, but faster.
- There's poor locking behavior around upsertEndpoint and elsewhere. I improved this now for the case of _this change_, but there are further cases that now I need to look into - including at least one I want to look at for this change.
- As above, I'm not sure plumbing through IsWireGuardOnly to the endpoint is necessarily a better change here. That would cover fewer cases, and I would end up wanting to do something like short-circuit on IsWireGuardOnly, but then fall down to checking discoShort or discoKey to correctly cover the case of older nodes without disco keys. Swapping the method out to be ep.isWireGuardOnly has more subtle usage throughout, which I don't think is a good approach - call sites need more expert knowledge about the difference between IsWireGuardOnly and pre-0.100 nodes. This method instead is specific and general around the state that needs protection.

If you have more specific concerns you can articulate, can you expand those?

Reply:
thanks for the info, no concerns was just asking for context

Reply:
After fixing up the synchronization, this method disappeared.